### ISSUE TITLE
Replacing loops with safer constructs

### ISSUE THREAD

**Issue by katrinleinweber** [TIBHannover/FAIR-studyGroup#3]
Constructing your own `for`-loops is only rarely actually necessary. We'll learn about alternatives here. Bring your own code and let's de-loop it! Good introduction in German: [Heise Developer: Weg mit den Schleifen!](https://www.heise.de/developer/artikel/Weg-mit-den-Schleifen-4009774.html?seite=all)

#### What should/will the [learning objective(s)](http://teachtogether.tech/en/process/#s:process-objectives) be?

- converting R `for`-loops to the [`apply()`](https://www.rdocumentation.org/packages/base/versions/3.5.1/topics/apply) [family](https://www.rdocumentation.org/packages/base/versions/3.5.1/topics/lapply) or [`purrr::map()`](https://purrr.tidyverse.org/reference/index.html)s
- similar concepts in Python (@david-morris)

#### What do participants need to know / install / bring?

- [RStudio](https://www.rstudio.com/products/rstudio/download/#download)
- either their own R code with `for` loops, or an example we'll provide
  - maybe this [RStudio](https://resources.rstudio.com/wistia-rstudio-conf-2017/happy-r-users-purrr-tutorial-charlotte-wickham) [example](https://github.com/rstudio/rstudio-conf/tree/master/2017/Happy_R_Users_Purrr-Charlotte_Wickham) ([backup](https://github.com/katrinleinweber/workshop-purrr))?

**Comment by jkreklow** [comment]
I have a couple of examples for de-looping in Python

**Comment by david-morris** [comment]
I could certainly help with this, and I'd be comfortable teaching functional Python after the teaching primer.  I've TA'd that class before...

But why?  Functional programming is honestly interesting mostly for people who are interested in code for code's sake.  Wouldn't our target audience be better off being told to use the `for item in collection` pattern and list comprehensions, which they would be more likely to use?

I think if we want to reduce the use of unpredictable structures, we should look at "a monad is a monoid in the category of endofunctors, what's the problem" as a _not_ do.

**Comment by jkreklow** [comment]
I agree that, in most cases, you'll use loops and list comprehensions. But as soon as it comes to time-consuming calculations on big datasets, I'm very happy to know how to use functional Python to save my day (and lots of time). So it probably depends on the audience we have...

**Comment by jkreklow** [comment]
Another very powerful tool for major efficiency improvements without much programming effort is just in time compilation with numba. This can be used to speed up pure numpy code with lots of loops instead of de-looping ;-)

**Comment by david-morris** [comment]
Something that meshes with efficiency and correctness concerns is if we teach people to manipulate data with the methods of the data structures they're using.  That's especially good when we're teaching people how to write code in languages with a global interpreter lock / green threads only, with data structures provided by libraries written in C++.

**Comment by katrinleinweber** [comment]
> But why? Functional programming is honestly interesting mostly for people who are interested in code for code's sake.

I take the opposite stand here: the functional paradigm avoids `for` loops as an error source and may prove more intuitive/natural. Maybe it's a brain-type-thingy ;-)

**Comment by david-morris** [comment]
I think that for loops are more universal between the languages scientists would use, and if we can get them to not use while loops, that would be an effective way of preventing  the errors mentioned on that blog recommending functional programming.

After all, the choice is between learning how to use for loops, or learning how to use map, reduce, zip, and select (they have to write functions anyway).  To someone who sees programming as a tax they have to pay to do science, I'd expect the first sounds more appealing, so they're going to do it, whether they have us around to tell us how to do it or not.

**Comment by david-morris** [comment]
Rereading my previous comments makes me think we could sell functional programming as "how to work with large datasets."

Come for the map-reduce and array broadcasts, stay for clean functional code.

**Comment by katrinleinweber** [comment]
Hi @david-morris! I'm not an expert in `list comprehension` in python. If they are comparable to R's `purrr::map()`, we can include them. Or make a separate event out of it?
